### ISSUE TITLE
[Grace CPU] How to Globally Override Compilers' Architecture Optimization Flags ? 

### ISSUE THREAD

**Issue by vitduck** [LLNL/spack#45293]
Hi, 

As mentions in #45227 , any compiler newer than `gcc@11` will fail during configure stage. 

1. Incorrect `-march`
    A common error message is generated during compilation of  a sample C code  
    ```
    configure:3930: checking whether the C compiler works
    configure:3952: /scratch/optpar01/spack/lib/spack/env/gcc/gcc    conftest.c  >&5
    Assembler messages:
    Error: unknown architecture `armv9-a+crc+profile+rng+memtag+sve2-bitperm+i8mm+bf16'
    Error: unrecognized option -march=armv9-a+crc+profile+rng+memtag+sve2-bitperm+i8mm+bf16
    ````
    I am puzzled from where the above `-march` was injected. 
   
2. Attempting to override via `compilers.yaml` 
    ``` 
    - compiler:
    spec: gcc@=13.3.0
    paths:
      cc: /scratch/optpar01/spack/opt/spack/linux-rocky9-neoverse_v2/gcc-11.4.1/gcc-13.3.0-r4yrnh2vuvwdlwxbjcbgxw7ztbufaadd/bin/gcc
      cxx: /scratch/optpar01/spack/opt/spack/linux-rocky9-neoverse_v2/gcc-11.4.1/gcc-13.3.0-r4yrnh2vuvwdlwxbjcbgxw7ztbufaadd/bin/g++
      f77: /scratch/optpar01/spack/opt/spack/linux-rocky9-neoverse_v2/gcc-11.4.1/gcc-13.3.0-r4yrnh2vuvwdlwxbjcbgxw7ztbufaadd/bin/gfortran
      fc: /scratch/optpar01/spack/opt/spack/linux-rocky9-neoverse_v2/gcc-11.4.1/gcc-13.3.0-r4yrnh2vuvwdlwxbjcbgxw7ztbufaadd/bin/gfortran
    flags: 
      cflags: -mcpu=neoverse-v2
      cppflags: -mcpu=neoverse-v2
    operating_system: rocky9
    target: aarch64
    modules: []
    environment: {}
    extra_rpaths: []
    ```
    However, this leads to another cryptic error message: 
    ```
    $ spack install numactl %gcc@13  
    Package numactl cannot satisfy '%gcc@13'
    ```

3.  Situation of Grace CPU 
![image](https://github.com/user-attachments/assets/ee01283f-0b98-4f32-bb37-96578d73b702)
The above slide is from Dr. John Linford, NVIDIA discussing the important of using correct `-mcpu` flag. 
So while it seems that `-march=armv9-a+crc+profile+rng+memtag+sve2-bitperm+i8mm+bf16` is equivalent to `-mcpu=neoverse-v2`, the compiler itself doesn't recognize the full extend of the instruction set. 

    Likewise, the opposite issue is `gcc@11.4.1` provided by RH 9.3 supports `-mcpu=neoverse-v2` via upstream patches, but I am stuck with `Armv8` optimization flag. 
 
    Hi @alalazo 
    Is there some simple way to override the flags globally ? 
    Thanks

**Comment by Candles** [comment]
Cross-posting from #45227, the issue is that GCC itself converts neoverse-v2 to armv9-a+crc+profile+rng+memtag+sve2-bitperm+i8mm+bf16, but Spack uses the system install of binutils by default and it doesn't support armv9-a on RHEL9 (and you can't easily override it since -march takes precedence over -mcpu).

You can fix it by recompiling gcc@13+binutils.

**Comment by vitduck** [comment]
@Candles 

It works now ! 
I don't know how to express my gratitude. 

Like you, I was haunted by this issue for weeks, and decided to settle with less optimized builds. 
I didn't learn of `binutils` until you mentioned it today ㅠㅠ

Many thanks.

**Comment by adamjstewart** [comment]
Seems like this issue was fixed, but feel free to reopen if that's not the case.

**Comment by lrbison** [comment]
I'm leaving this note for anyone else who encounters this issue on neoverse-v2 (Graviton4 or Grace)

Based on the above thread I was able to confirm spack's gcc@13.2.0 was using /usr/bin/as with `-v`:

```
gcc -c -v test.c -mcpu=neoverse-v2
...
COLLECT_GCC_OPTIONS='-v' '-c' '-mcpu=neoverse-v2' '-mlittle-endian' '-mabi=lp64' '-B' '/usr/bin' \
/usr/bin/as -v -EL -march=armv9-a+crc+profile+rng+memtag+sve2-bitperm+i8mm+bf16 -mabi=lp64 -o noop.o /tmp/ccUcL8KN.s
```

After re-installing gcc@13.2+binutils, the situation remained unchanged, and I was still calling `/usr/bin/as`.  The reason was that my `packages.yaml` already had an (older) external binutils listed, and so spack happily re-used that one when satisfying `+binutils`

So in short, you may need to be explicit about using a newer binutils, something like `gcc@13.2.0+binutils ^binutils@2.42`
